fix(@desktop/sugnup): Profile image chosen during signup sould be saved for the user and squish test for the same

Khushboo-dev-cpp opened the pull request:
fix(@desktop/sugnup): Profile image chosen during signup sould be saved for the user and squish test for the same

fixes #6861

### What does the PR do

Fixes the issue of the profile image not stores after setting it during the signup process.
Also added squish test for the use case.

### Affected areas

Signup, Profile

### Screenshot of functionality (including design for comparison)

- [x] I've checked the design and this PR matches it

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 8b0ae401 | [#1](https://ci.status.im/job/status-desktop/job/branches/job/macos/job/PR-7075/1/) | 2022-08-18 07:50:07 | ~8 min | `macos` | [:package:`dmg`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Desktop-220818-074131-8b0ae4-pr7075.dmg) |

jrainville commented on test/ui-test/src/screens/StatusWelcomeScreen.py:
```diff
@@ -114,3 +125,39 @@ def _agree_terms_and_conditions(self):
+        # take a screenshot of the profile image to compare it later with the main screen
+        profileIcon = wait_and_get_obj(SignUpComponents.WLECOME_SCREEN_USER_PROFILE_IMAGE.value)
+        img = object.grabScreenshot(profileIcon, {"delay": 0})
+        img.save("../shared/searchImages/profiletestimage.png")
```
Yo nice.

jrainville commented on test/ui-test/testSuites/suite_status/shared/scripts/bdd_hooks.py:
```diff
@@ -16,6 +16,7 @@ def hook(context):
     context.userData = {}
     context.userData["aut_name"] = _status_desktop_app_name
     context.userData["status_data_folder_path"] = _status_data_folder_path
+    context.userData["fixtures_root"] = os.path.join(os.path.dirname(__file__), "../../../fixtures/")
```
FYI, we found that this doesn't work for Mac. I updated my PR (where you most likely took inspiration :wink: ) so you can update with that.

Khushboo-dev-cpp replied:
hehe ok, close save :D

jrainville commented on ui/imports/shared/popups/ImageCropWorkflow.qml:
```diff
@@ -26,6 +27,11 @@ Item {
         fileDialog.open()
     }
 
+    function cropImage(imageUrl) {
```
This function lets you skip the file dialog right? If so, then my comment on the fixtures_root is irrelevant. I might copy you for my PR in the end.

Khushboo-dev-cpp replied:
aaha yes, its does skip the filedialog completely!

jrainville reviewed (commented):
Very good job!
Super cool that you make the snapshot feature work. Do you know if it cleans up the image after? Or do we even need to?

Otherwise, I have not approved yet just because I want to check about the fixtures_root code first. 

Also, bonus points next time if you split the PR in two commits, one for the fix, the other for the test :smile:

Khushboo-dev-cpp commented:
> Very good job! Super cool that you make the snapshot feature work. Do you know if it cleans up the image after? Or do we even need to?
> 
> Otherwise, I have not approved yet just because I want to check about the fixtures_root code first.
> 
> Also, bonus points next time if you split the PR in two commits, one for the fix, the other for the test 😄

hmmm, good point about the image being deleted, let me verify

Khushboo-dev-cpp commented:
@jrainville @noelia-santos this doesnt work like I expected, its weird cause it did yesterday, so I am debugging further. just fyi!

noeliaSD commented on test/ui-test/src/screens/StatusWelcomeScreen.py:
```diff
@@ -114,3 +133,55 @@ def _agree_terms_and_conditions(self):
     def seed_phrase_visible(self):
         is_loaded_visible_and_enabled(SeedPhraseComponents.INVALID_SEED_TEXT.value)
         
+    def input_profileImage(self, profileImageUrl: str):
```
```suggestion
    def input_profile_image(self, profileImageUrl: str):
```

noeliaSD commented on test/ui-test/src/screens/StatusWelcomeScreen.py:
```diff
@@ -47,6 +49,22 @@ class SeedPhraseComponents(Enum):
+    
+class MainScreen(Enum):
+    PROFILE_NAV_BAR_BUTTON: str = "mainWindow_ProfileNavBarButton"
+    SETTINGS_BUTTON = "settings_navbar_settings_icon_StatusIcon"
```
Here, instead of creating the `MainScreen` enum, you can import the existing one that already contains the settings button (import from `StatusMainScreen`, `MainScreenComponents`).

noeliaSD commented on test/ui-test/src/screens/StatusWelcomeScreen.py:
```diff
@@ -114,3 +133,55 @@ def _agree_terms_and_conditions(self):
+        
+        # take a screenshot of the profile image to compare it later with the main screen
+        profileIcon = wait_and_get_obj(SignUpComponents.WELCOME_SCREEN_USER_PROFILE_IMAGE.value)
+        img = object.grabScreenshot(profileIcon, {"delay": 200})
```
Use `SquishDriver` instead of this direct call.

noeliaSD commented on test/ui-test/src/screens/StatusWelcomeScreen.py:
```diff
@@ -114,3 +133,55 @@ def _agree_terms_and_conditions(self):
+    def profile_modal_image_is_updated(self):
+        click_obj_by_name(MainScreen.PROFILE_NAV_BAR_BUTTON.value)
+        click_obj_by_name(UserContextStatusMenu.USER_CONTEXT_MENU_VIEW_MY_PROFILE_BUTTON.value)
+        test.imagePresent("profiletestimage", {"tolerant": True, "threshold": 97, "minScale": 95, "maxScale": 100, "multiscale": True })
```
Use `SquishVerificationDriver` for this validation.

noeliaSD commented on test/ui-test/src/screens/StatusWelcomeScreen.py:
```diff
@@ -114,3 +133,55 @@ def _agree_terms_and_conditions(self):
+         # first time clicking on settings button closes the my profile modal
+        click_obj_by_name(MainScreen.SETTINGS_BUTTON.value)
+        click_obj_by_name(MainScreen.SETTINGS_BUTTON.value)
+        test.imagePresent("profiletestimage", { "tolerant": True, "threshold": 97, "minScale": 100, "maxScale": 183, "multiscale": True })
```
Same than before! Validation wrapper

noeliaSD commented on test/ui-test/src/screens/StatusWelcomeScreen.py:
```diff
@@ -114,3 +133,55 @@ def _agree_terms_and_conditions(self):
+    def grab_screenshot(self):
+                # take a screenshot of the profile image to compare it later with the main screen
+        loginUserName = wait_and_get_obj(LoginView.LOGIN_VIEW_USER_IMAGE.value)
+        img = object.grabScreenshot(loginUserName, {"delay": 200})
```
Same than before driver wrapper.

noeliaSD commented on test/ui-test/src/screens/StatusWelcomeScreen.py:
```diff
@@ -114,3 +133,55 @@ def _agree_terms_and_conditions(self):
+        
+    def profile_image_is_updated(self):
+        profileNavBarButton = wait_and_get_obj(MainScreen.PROFILE_NAV_BAR_BUTTON.value)
+        test.imagePresent("loginUserName", { "tolerant": True, "threshold": 98, "minScale": 85, "maxScale": 100, "multiscale": True })
```
Wrapper

noeliaSD commented on test/ui-test/src/screens/StatusWelcomeScreen.py:
```diff
@@ -114,3 +133,55 @@ def _agree_terms_and_conditions(self):
+        test.imagePresent("loginUserName", { "tolerant": True, "threshold": 98, "minScale": 85, "maxScale": 100, "multiscale": True })
+
+        click_obj(profileNavBarButton)
+        test.imagePresent("loginUserName", { "tolerant": True, "threshold": 99, "minScale": 95, "maxScale": 100, "multiscale": True })
```
Wrapper

noeliaSD commented on test/ui-test/src/utils/FileManager.py:
```diff
@@ -1,8 +1,10 @@
 import shutil
 import distutils.dir_util
-
+import test
```
Maybe better to continue with the pattern by using our wrapper

noeliaSD commented on test/ui-test/testSuites/suite_status/shared/steps/signUpSteps.py:
```diff
@@ -41,3 +41,26 @@ def step(context, seed_phrase):
 def step(context):
     _welcomeScreen.seed_phrase_visible()
+    
+@When("the user signs up with profileImage |any| and username |any| and password |any|")
```
```suggestion
@When("the user signs up with profileImage |any|, username |any| and password |any|")
```

noeliaSD commented on test/ui-test/testSuites/suite_status/shared/steps/signUpSteps.py:
```diff
@@ -41,3 +41,26 @@ def step(context, seed_phrase):
+def step(context, profileImageUrl, username, password):
+    _welcomeScreen.input_username_profileImage_password_and_finalize_sign_up("file:///"+context.userData["fixtures_root"]+"images/"+profileImageUrl, username, password)
+ 
+@Then("the my profile modal has the updated profile image")
```
```suggestion
@Then("my profile modal has the updated profile image")
```

noeliaSD commented on test/ui-test/testSuites/suite_status/shared/steps/signUpSteps.py:
```diff
@@ -41,3 +41,26 @@ def step(context, seed_phrase):
+def step(context):
+    _welcomeScreen.grab_screenshot()
+ 
+@Then("the profile navBar has the updated profile image")
```
```suggestion
@Then("the profile navigation bar has the updated profile image")
```

noeliaSD commented on test/ui-test/testSuites/suite_status/shared/steps/signUpSteps.py:
```diff
@@ -41,3 +41,26 @@ def step(context, seed_phrase):
+def step(context):
+    _welcomeScreen.profile_modal_image_is_updated()
+    
+@Then("profile settings has the updated profile image")
```
??
```suggestion
@Then("the profile setting has the updated profile image")
```

noeliaSD reviewed (changes requested):
Great job!!!

Added some minor comments and I share a video with the failure I'm experiencing (maybe bc my OS is `mac`:

https://user-images.githubusercontent.com/97019400/185857793-1a25b93e-3a45-40df-a896-db3d3776d9ff.mov

Khushboo-dev-cpp commented:
@noeliaSD @jrainville I have addressed all review comments. Noelia please check if it now works on macOS, I fixed the error you were facing but not sure if there will be other path related issues

noeliaSD reviewed (approved):
Super cool test!! 

Please review my comments added but already approved (tested and now works fine)!

Khushboo-dev-cpp commented on test/ui-test/src/drivers/SquishDriver.py:
```diff
@@ -28,6 +28,7 @@
 # The default minimum timeout to find ui object
 _MIN_WAIT_OBJ_TIMEOUT = 500
 
+_SEARCH_IMAGES_PATH = "../shared/searchImages/"
```
he its not the same path, this is the path to store searchImages and that was the profile image test data.

Khushboo-dev-cpp commented on test/ui-test/testSuites/suite_status/shared/scripts/sections/login_names.py:
```diff
@@ -7,3 +7,4 @@
 loginView_main = {"container": statusDesktop_mainWindow, "type": "LoginView", "visible": True}
 loginView_errMsgLabel = {"container": statusDesktop_mainWindow, "objectName": "loginPassworkInputValidationErrorText", "type": "StatusBaseText", "visible": True}
 accountsView_accountListPanel = {"container": statusDesktop_mainWindow, "type": "ListView", "visible": True} # This probably is missing an objectName
+loginView_userImage = {"container": statusDesktop_mainWindow, "objectName": "loginViewUserImage", "type": "UserImage", "visible": True}
```
aaah yes it should!

Khushboo-dev-cpp commented on test/ui-test/testSuites/global_shared/scripts/global_names.py:
```diff
@@ -18,4 +22,4 @@
 join_public_chat_StatusMenuItemDelegate = {"checkable": False, "container": statusDesktop_mainWindow_overlay, "enabled": True, "text": "Join public chat", "type": "StatusMenuItemDelegate", "unnamed": 1, "visible": True}
-mainWindow_statusIcon_StatusIcon_2 = {"container": statusDesktop_mainWindow, "id": "statusIcon", "source": "qrc:/StatusQ/src/assets/img/icons/public-chat.svg", "type": "StatusIcon", "unnamed": 1, "visible": True}
\ No newline at end of file
+mainWindow_statusIcon_StatusIcon_2 = {"container": statusDesktop_mainWindow, "id": "statusIcon", "source": "qrc:/StatusQ/src/assets/img/icons/public-chat.svg", "type": "StatusIcon", "unnamed": 1, "visible": True}
```
apparently was not used anywhere. removed!
